### PR TITLE
Update ensign.py

### LINKED ISSUE #69: Type annotations
**Describe the solution you'd like**
In `ensign.py` we would like to use type annotations in the function definitions so that it's clear what expected types the parameters should have.

### PR DESCRIPTION
**Pull request by Dv04**
This PR fixes #69 _(If you are fixing a bug)_ which reported a bug that caused a problem to occur when users...

I have made the following changes:

1. Added Types to all the needed places
2. Formatted the code a bit
3. Added types from typing module

### CHECKLIST

- [x] _Is the commit message formatted correctly?_

- [x] _Do all of your functions and methods have docstrings?_
- [x] _Have you added/updated unit tests where appropriate?_
- [ ] _Have you run the unit tests using `pytest`?_
- [x] _Is your code style correct (are you using PEP8, pyflakes)?_

### CODE REVIEW

File: pyensign/ensign.py [hunk 1]
```diff
@@ -28,6 +28,21 @@
...
+    Union,
+    AsyncGenerator,
+    Dict,
+    Any,
+    Iterable,
+    Coroutine,
+    Type,
+)
+
+from typing import Any, Callable
```
Comment (pdeziel) [inline pyensign/ensign.py:44]: Can we import `Any` and `Callable` in the same group above?

File: pyensign/ensign.py [hunk 2]
```diff
@@ -107,8 +122,13 @@ def __init__(
         self.client = Client(connection, topic_cache=self.topics)
 
     async def publish(
-        self, topic, *events, on_ack=None, on_nack=None, ensure_exists=False
-    ):
+        self,
+        topic: str,
+        *events: Any,  # This is inferred to be any type, but might be narrowed down later.
```
Comment (pdeziel) [inline pyensign/ensign.py:127]: Can we remove the comment? I'd like to keep comments in the docstring so the function parameters are concise and readable. The `Any` type hint is a common thing that people can look up so I don't think we are obligated to explain that here.

File: pyensign/ensign.py [hunk 3]
```diff
@@ -195,7 +215,13 @@ async def next():
             on_nack=on_nack,
         )
 
-    async def subscribe(self, *topics, query="", params=None, consumer_group=None):
+    async def subscribe(
+        self,
+        *topics: str,
+        query: str = "",
+        params: Optional[Dict[str, Any]] = None,
+        consumer_group: Optional[Any] = None  # Placeholder for actual type
```
Comment (pdeziel) [inline pyensign/ensign.py:223]:
```suggestion
        consumer_group: Optional[Any] = None
```

File: pyensign/ensign.py [hunk 4]
```diff
@@ -323,7 +351,7 @@ async def query(self, query, params=None):
     async def explain_query(self, query, params):
         raise NotImplementedError
 
-    async def get_topics(self):
+    async def get_topics(self) -> List[Any]:  # Placeholder for actual list type
```
Comment (pdeziel) [inline pyensign/ensign.py:354]:
```suggestion
    async def get_topics(self) -> List[Any]:
```

File: pyensign/ensign.py (hunk 2 above)
Comment (pdeziel) [reply pyensign/ensign.py:127]: Actually if possible this should be the custom `Event` type here instead of `Any`.

File: pyensign/ensign.py [hunk 5]
```diff
@@ -677,8 +708,10 @@ def _resolve_topic(self, topic):
 # asyncio coroutines.
 _client = None
 
-
-def authenticate(*auth_args, **auth_kwargs):
+def authenticate(
+    *auth_args: Any, 
+    **auth_kwargs: Any
+) -> Callable[[Union[Coroutine[Any, Any, Any], AsyncGenerator[Any, None]]], Union[Coroutine[Any, Any, Any], AsyncGenerator[Any, None]]]:
```
Comment (pdeziel) [inline pyensign/ensign.py:714]: Thanks for being thorough with the annotations!

**Review by pdeziel — CHANGES_REQUESTED** [review]
This is coming along great, I had some very small suggestions and we also have to make sure that the code is formatted consistently with the rest of the codebase.

To get the CI to pass, can you setup the pre-commit hooks on your end?

```bash
pip install -r requirements.txt
pre-commit install
```

Then on your next commit it should reformat the file automatically for you (you may have to `git add` and `git commit` again once the reformat happens).

File: pyensign/ensign.py (hunk 1 above)
Comment (Dv04) [reply pyensign/ensign.py:44]: Alright

**Review by pdeziel — APPROVED** [review]
Nice work! Thanks for making the changes and the annotations look great!

**Pull request merged**